File: javascript/src/infraestrutura/ArmazenamentoS3.js
import { DocumentoUploadPort } from '../porta/DocumentoUploadPort.js';
import { DocumentoDownloadPort } from '../porta/DocumentoDownloadPort.js';
import { implementa } from '../porta/_contrato.js';

/**
 * ISP na pratica: um unico adaptador cumpre DUAS interfaces pequenas.
 * Como JS so tem heranca simples, usamos o helper implementa() para compor os
 * dois contratos — o equivalente ao "implements A, B" do Java.
 * Quem so envia depende de DocumentoUploadPort e nem sabe que baixar existe.
 */
export class ArmazenamentoS3 extends implementa(DocumentoUploadPort, DocumentoDownloadPort) {
  #bucket;
  #regiao;
  #objetos = new Map();

  constructor(bucket, regiao) {
    super();
    this.#bucket = bucket;
    this.#regiao = regiao;
  }

  enviar(documento) {
    this.#objetos.set(documento.id, documento);
    console.log(`[s3://${this.#bucket}.${this.#regiao}] upload de ${documento}`);
    return documento.id;
  }

  baixar(id) {
    const doc = this.#objetos.get(id);
    if (!doc) {
      throw new Error(`Documento inexistente: ${id}`);
    }
    return doc;
  }
}

File: java/src/br/unifor/facilitador/infraestrutura/ArmazenamentoS3.java
package br.unifor.facilitador.infraestrutura;

import br.unifor.facilitador.dominio.Documento;
import br.unifor.facilitador.porta.DocumentoDownloadPort;
import br.unifor.facilitador.porta.DocumentoUploadPort;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

/**
 * ISP na pratica: um unico adaptador implementa DUAS interfaces pequenas.
 * Quem so envia depende de DocumentoUploadPort e nem sabe que baixar existe.
 */
public class ArmazenamentoS3 implements DocumentoUploadPort, DocumentoDownloadPort {

    private final String bucket;
    private final String regiao;
    private final Map<UUID, Documento> objetos = new HashMap<>();

    public ArmazenamentoS3(String bucket, String regiao) {
        this.bucket = bucket;
        this.regiao = regiao;
    }

    @Override
    public UUID enviar(Documento d) {
        objetos.put(d.getId(), d);
        System.out.println("[s3://" + bucket + "." + regiao + "] upload de " + d);
        return d.getId();
    }

    @Override
    public Documento baixar(UUID id) {
        Documento d = objetos.get(id);
        if (d == null) {
            throw new IllegalArgumentException("Documento inexistente: " + id);
        }
        return d;
    }
}
